test/e2e/metering_manual_install_test.go

```diff
@@ -107,6 +110,111 @@ func testManualMeteringInstall(
 	assert.NoError(t, err, "capturing logs and uninstalling metering should produce no error")
 }
 
+func createNFSProvisioner(ctx *deployframework.DeployerCtx) error {
+	//Waiting for Metering pods
+	ctx.TargetPodsCount = nonHDFSTargetPodCount
+
+	_, err := createTestingNamespace(ctx.Client, nfsTestingNamespace)
+	if err != nil {
+		return err
+	}
+
+	files := map[string]string{
+		"pod":          "server.yaml",
+		"service":      "service.yaml",
+		"storageClass": "storageclass.yaml",
+	}
+	server := &corev1.Pod{}
+	service := &corev1.Service{}
+	storageClass := &storagev1beta1.StorageClass{}
+	for name, file := range files {
+		absFile := filepath.Join(repoPath, testNFSManifestPath, file)
+		switch name {
+		case "pod":
+			if err := deploy.DecodeYAMLManifestToObject(absFile, server); err != nil {
+				return err
+			}
+			ctx.Client.CoreV1().Pods(nfsTestingNamespace).Create(context.TODO(), server, metav1.CreateOptions{})
+			if err != nil && !apierrors.IsAlreadyExists(err) {
+				return err
+			}
+		case "service":
+			if err := deploy.DecodeYAMLManifestToObject(absFile, service); err != nil {
+				return err
+			}
+			ctx.Client.CoreV1().Services(nfsTestingNamespace).Create(context.TODO(), service, metav1.CreateOptions{})
+			if err != nil && !apierrors.IsAlreadyExists(err) {
+				return err
+			}
+		case "storageClass":
+			if err := deploy.DecodeYAMLManifestToObject(absFile, storageClass); err != nil {
+				return err
+			}
+			ctx.Client.StorageV1beta1().StorageClasses().Create(context.TODO(), storageClass, metav1.CreateOptions{})
+			if err != nil && !apierrors.IsAlreadyExists(err) {
+				return err
+			}
+		}
+	}
+
+	svc, err := ctx.Client.CoreV1().Services(nfsTestingNamespace).Get(context.TODO(), "nfs-service", metav1.GetOptions{})
+	if apierrors.IsNotFound(err) {
+		return err
+	}
+	if err != nil || svc.Spec.ClusterIP == "" {
+		err = wait.Poll(3*time.Second, 5*time.Minute, func() (bool, error) {
+			svc, err = ctx.Client.CoreV1().Services(ctx.Namespace).Get(context.TODO(), "nfs-service", metav1.GetOptions{})
+			if apierrors.IsNotFound(err) {
+				return false, err
+			}
+			if err != nil {
+				return false, nil
+			}
+			return svc.Spec.ClusterIP != "" , nil
+		})
+		if err != nil {
+			return err
+		}
+	}
+
+	nfsFile := "persistentvolume.yaml"
+	nfsPv := &corev1.PersistentVolume{}
+	absFile := filepath.Join(repoPath, testNFSManifestPath, nfsFile)
+	//decode pvc
+	if err := deploy.DecodeYAMLManifestToObject(absFile, nfsPv); err != nil {
+		return err
+	}
+
+	nfsPv.Spec.NFS.Server = svc.Spec.ClusterIP
+	_, err = ctx.Client.CoreV1().PersistentVolumes().Create(context.TODO(), nfsPv, metav1.CreateOptions{})
+	if err != nil && !apierrors.IsAlreadyExists(err) {
+		return err
+	}
+	return nil
```

Comment:
As a follow-up, we should aim towards making this more idempotent.

Reply:
Aside from the test output dir fail thing, wouldn't running this test over and over have the same results and effects?

Reply:
I realize the code block I'm highlighting here is for the PV setup, and idempotent was a poor choice of words - we can further refine this implementation by checking if these resources already exist, and if not, attempt to create the resource. I don't think we should care about the upgrade scenario anyways as PV's are immutable post creation and that just introduces extra work when tearing down the namespace is sufficient (which we do when exiting the e2e suite in non-dev mode) in most cases.